pkg/machines/libvirt.es6

```diff
@@ -218,12 +223,79 @@ LIBVIRT_PROVIDER = {
                 dispatch(checkLibvirtStatus(libvirtServiceName));
                 startEventMonitor(dispatch, connectionName, libvirtServiceName);
                 doGetAllVms(dispatch, connectionName);
+                dispatch(getStoragePools(connectionName));
             };
         }
 
         return unknownConnectionName(getAllVms, libvirtServiceName);
     },
 
+    /**
+     * Retrieves list of libvirt "storage pools" for particular connection.
+     */
+    GET_STORAGE_POOLS({ connectionName }) {
+        logDebug(`${this.name}.GET_STORAGE_POOLS(connectionName='${connectionName}')`);
+
+        const connection = VMS_CONFIG.Virsh.connections[connectionName].params.join(' ');
+        // Workaround: virsh v1.3.1 in ubuntu-1604 does not support '--name' parameter
+        const command = `virsh ${connection} -r pool-list --type dir | grep active | awk '{print $1 }'`;
+        let poolList = '';
+        // TODO: add support for other pool types then just the "directory"
+        return dispatch => cockpit
+                .script(command, null, { err: "message", environ: ['LC_ALL=en_US.UTF-8'] })
+                .stream(output => { poolList += output; })
+                .then(() => { // so far only pool names are needed, extend here otherwise
+                    const promises = parseStoragePoolList(dispatch, connectionName, poolList)
+                            .map(poolName => dispatch(getStorageVolumes(connectionName, poolName)));
+
+                    return cockpit.all(promises);
+                })
+                .fail((exception, data) => {
+                    console.error('Failed to get list of Libvirt storage pools for connection ', connectionName, ': ', data, exception);
+                });
+    },
+
+    GET_STORAGE_VOLUMES({ connectionName, poolName }) {
+        logDebug(`${this.name}.GET_STORAGE_VOLUMES(connectionName='${connectionName}', poolName='${poolName}')`);
+        const connection = VMS_CONFIG.Virsh.connections[connectionName].params.join(' ');
+        // Caution: output of virsh v1.3.1 (ubuntu-1604) and v3.7.0 differs, the 'grep' unifies it
+        const command = `virsh ${connection} -q pool-refresh ${poolName} && virsh ${connection} -q -r vol-list ${poolName} --details | (grep file || true)`;
```

Comment:
Can pool names contain spaces, quotes, or other shell-specific characters?

Reply:
Good catch, the character set is same as for file names.

Same issue for VM names and others which are recently not handled either.

Since it's a generic bug, I'll open an issue for that.
Are arguments already escaped somewhere within other parts of Cockpit? I found a few npm packages wrapping regular expressions to deal with thatt, but this should be consistent.

Reply:
Not with `cockpit.script()`, that takes a verbatim shell string and executes it as-is. You can do this more robustly with `cockpit.spawn()` and passing a command array, but of coures then `&&` does not work, you need two spawns and chain them.

Reply:
No matter how serious this issue might be, it is not originally introduced by this PR. Can it be fixed within #9623 ? So merging of this and stacked PRs will not be blocked.

Reply:
How is it not introduced here? The pool-refresh and similar commands (`virsh.*pool`) don't exist yet in master. But OK for me to fix in a follow-up.

Reply:
It's same issue as for VM names or other objects. This PR keeps repeating the bug.